### PR TITLE
Add function to invalidate nonce

### LINKED ISSUE #66: Implement an external method to allow the caller to invalidate nonce
This method would simply use the nonce of the msg.sender. Its purpose would be to invalidate a previously signed message with the most recent nonce. The method should also be unit tested.

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Add function that allows a user to invalidate their nonce

Closes #66

### CODE REVIEW

File: test/UniStaker.t.sol [hunk 1]
```diff
@@ -2625,6 +2625,72 @@ contract SetAdmin is UniStakerTest {
...
+
+  function testFuzz_NonOwnerCannotUpdateAnotherAddressesNonce(
+    address _caller,
+    address _nonceOwner,
+    uint256 _initialNonce
+  ) public {
+    vm.assume(_caller != address(0) && _caller != _nonceOwner);
+    vm.assume(_initialNonce != type(uint256).max);
+
+    // Set nonces in storage assert incremented by one
```
Comment (wildmolasses) [inline test/UniStaker.t.sol:2658]: nit: can remove comment

File: test/UniStaker.t.sol [hunk 2]
```diff
@@ -2625,6 +2625,72 @@ contract SetAdmin is UniStakerTest {
...
+    assertEq(currentNonce, _initialNonce);
+  }
+
+  function testFuzz_NonceIncrementsWhenCalledMultipleTimes(address _caller, uint256 _initialNonce)
+    public
+  {
+    vm.assume(_caller != address(0));
+    _initialNonce = bound(_initialNonce, 0, type(uint256).max - 2);
+
+    // Set nonces in storage assert incremented by one
```
Comment (wildmolasses) [inline test/UniStaker.t.sol:2677]: think we can remove comment

**Review by wildmolasses — COMMENTED** [review]
some small comments, otherwise lgtm

File: test/UniStaker.t.sol [hunk 3]
```diff
@@ -2625,6 +2625,72 @@ contract SetAdmin is UniStakerTest {
...
+
+    vm.prank(_caller);
+    uniStaker.invalidateNonce();
+
+    uint256 currentNonce = uniStaker.nonces(_caller);
+
+    assertEq(currentNonce, _initialNonce + 1, "Current nonce is incorrect");
+  }
+
+  function testFuzz_NonOwnerCannotUpdateAnotherAddressesNonce(
```
[the remark this answers is not in the document]
Comment (alexkeating) [reply test/UniStaker.t.sol:2673]: Yes, I think that is fair

File: test/UniStaker.t.sol (hunk 2 above)
Comment (alexkeating) [reply test/UniStaker.t.sol]: https://github.com/uniswapfoundation/UniStaker/pull/72/commits/fe9d1c612c45e2b0701f3117d91da7d17704617b

File: test/UniStaker.t.sol (hunk 1 above)
Comment (alexkeating) [reply test/UniStaker.t.sol]: https://github.com/uniswapfoundation/UniStaker/pull/72/commits/fe9d1c612c45e2b0701f3117d91da7d17704617b

File: test/UniStaker.t.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexkeating) [reply test/UniStaker.t.sol]: https://github.com/uniswapfoundation/UniStaker/pull/72/commits/fe9d1c612c45e2b0701f3117d91da7d17704617b

File: test/UniStaker.t.sol [hunk 4]
```diff
@@ -2625,6 +2625,72 @@ contract SetAdmin is UniStakerTest {
...
+contract InvalidateNonce is UniStakerTest {
+  using stdStorage for StdStorage;
+
+  function testFuzz_SucessfullyInvalidatesNonceForMsgSender(address _caller, uint256 _initialNonce)
+    public
+  {
+    vm.assume(_caller != address(0));
+    vm.assume(_initialNonce != type(uint256).max);
+
+    // Set nonces in storage assert incremented by one
```
[the remark this answers is not in the document]
Comment (alexkeating) [reply test/UniStaker.t.sol]: https://github.com/uniswapfoundation/UniStaker/pull/72/commits/fe9d1c612c45e2b0701f3117d91da7d17704617b

File: test/UniStaker.t.sol [hunk 5]
```diff
@@ -2648,6 +2648,49 @@ contract SetAdmin is UniStakerTest {
   }
 }
 
+contract InvalidateNonce is UniStakerTest {
+  using stdStorage for StdStorage;
+
+  function testFuzz_SucessfullyInvalidatesNonceForMsgSender(address _caller, uint256 _initialNonce)
```
Comment (apbendi) [inline test/UniStaker.t.sol:2654]: nit: test name could be a little clearer from the `scopelint spec` perspective. Maybe "Increments the nonce value of the sender"

File: test/UniStaker.t.sol [hunk 6]
```diff
@@ -2648,6 +2648,49 @@ contract SetAdmin is UniStakerTest {
...
+
+    vm.prank(_caller);
+    uniStaker.invalidateNonce();
+
+    uint256 currentNonce = uniStaker.nonces(_caller);
+
+    assertEq(currentNonce, _initialNonce + 1, "Current nonce is incorrect");
+  }
+
+  function testFuzz_NonceIncrementsWhenCalledMultipleTimes(address _caller, uint256 _initialNonce)
```
Comment (apbendi) [inline test/UniStaker.t.sol:2672]: similarly, perhaps: "Increases the nonce by 2 when called twice"

**Review by apbendi — APPROVED** [review]
Test naming nits, but otherwise looks good @alexkeating

File: test/UniStaker.t.sol (hunk 6 above)
Comment (alexkeating) [reply test/UniStaker.t.sol]: https://github.com/uniswapfoundation/UniStaker/pull/72/commits/335abf9be099488ad2a22ae173fe4546a0592762

**Pull request merged**